core/src/main/java/org/apache/iceberg/avro/Avro.java

```diff
@@ -143,24 +171,378 @@ public WriteBuilder overwrite(boolean enabled) {
       return this;
     }
 
-    private CodecFactory codec() {
-      String codec = config.getOrDefault(AVRO_COMPRESSION, AVRO_COMPRESSION_DEFAULT);
-      try {
-        return CodecName.valueOf(codec.toUpperCase(Locale.ENGLISH)).get();
-      } catch (IllegalArgumentException e) {
-        throw new IllegalArgumentException("Unsupported compression codec: " + codec);
-      }
+    // supposed to always be a private method used strictly by data and delete write builders
+    private WriteBuilder createContextFunc(Function<Map<String, String>, Context> newCreateContextFunc) {
+      this.createContextFunc = newCreateContextFunc;
+      return this;
     }
 
     public <D> FileAppender<D> build() throws IOException {
       Preconditions.checkNotNull(schema, "Schema is required");
       Preconditions.checkNotNull(name, "Table name is required and cannot be null");
 
+      Function<Schema, DatumWriter<?>> writerFunc;
+      if (createWriterFunc != null) {
+        writerFunc = createWriterFunc;
+      } else {
+        writerFunc = GenericAvroWriter::new;
+      }
+
       // add the Iceberg schema to keyValueMetadata
       meta("iceberg.schema", SchemaParser.toJson(schema));
 
+      Context context = createContextFunc.apply(config);
+      CodecFactory codec = context.codec();
+
       return new AvroFileAppender<>(
-          AvroSchemaUtil.convert(schema, name), file, createWriterFunc, codec(), metadata, overwrite);
+          schema, AvroSchemaUtil.convert(schema, name), file, writerFunc, codec, metadata, metricsConfig, overwrite);
+    }
+
+    private static class Context {
+      private final CodecFactory codec;
+
+      private Context(CodecFactory codec) {
+        this.codec = codec;
+      }
+
+      static Context dataContext(Map<String, String> config) {
+        String codecAsString = config.getOrDefault(AVRO_COMPRESSION, AVRO_COMPRESSION_DEFAULT);
+        CodecFactory codec = toCodec(codecAsString);
+
+        return new Context(codec);
+      }
+
+      static Context deleteContext(Map<String, String> config) {
+        // default delete config using data config
+        Context dataContext = dataContext(config);
+
+        String codecAsString = config.get(DELETE_AVRO_COMPRESSION);
+        CodecFactory codec = codecAsString != null ? toCodec(codecAsString) : dataContext.codec();
+
+        return new Context(codec);
+      }
+
+      private static CodecFactory toCodec(String codecAsString) {
+        try {
+          return CodecName.valueOf(codecAsString.toUpperCase(Locale.ENGLISH)).get();
+        } catch (IllegalArgumentException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)